src/agentscope/service/execute_code/exec_notebook.py

```diff
@@ -0,0 +1,198 @@
+            await self.nb_client.km.cleanup_resources()
+
+        self.nb_client.kc.stop_channels()
+        self.nb_client.kc = None
+        self.nb_client.km = None
+
+    async def _restart_client(self) -> None:
+        """Restart the notebook client"""
+        await self._kill_client()
+        self.nb_client = NotebookClient(self.nb, timeout=self.timeout)
+        await self._start_client()
+
+    async def _run_cell(self, cell_index: int) -> ServiceResponse:
+        """Run a cell in the notebook by its index"""
+        try:
+            self.nb_client.execute_cell(self.nb.cells[cell_index], cell_index)
+            return ServiceResponse(
+                status=ServiceExecStatus.SUCCESS,
+                content=[
+                    self.output_parser(output)
+                    for output in self.nb.cells[cell_index].outputs
+                ],
+            )
+        except DeadKernelError:
+            await self.reset()
+            return ServiceResponse(
+                status=ServiceExecStatus.ERROR,
+                content="DeadKernelError when executing cell, reset kernel",
+            )
+        except CellTimeoutError:
+            assert self.nb_client.km is not None
+            await self.nb_client.km.interrupt_kernel()
+            return ServiceResponse(
+                status=ServiceExecStatus.ERROR,
+                content=(
+                    "CellTimeoutError when executing cell"
+                    ", code execution timeout"
+                ),
+            )
+        except Exception as e:
+            return ServiceResponse(
+                status=ServiceExecStatus.ERROR,
+                content=str(e),
+            )
+
+    @property
+    def cells_length(self) -> int:
+        """return cell length"""
+        return len(self.nb.cells)
+
+    async def async_run_code_on_notebook(self, code: str) -> ServiceResponse:
+        """
+        Run the code on interactive notebook
+        """
+        self.nb.cells.append(nbformat.v4.new_code_cell(code))
+        cell_index = self.cells_length - 1
+        return await self._run_cell(cell_index)
+
+    def run_code_on_notebook(self, code: str) -> ServiceResponse:
+        """
+        Run the code on interactive jupyter notebook.
+
+        Args:
+            code (`str`):
+                The Python code to be executed in the interactive notebook.
+
+        Returns:
+            `ServiceResponse`: whether the code execution was successful,
+            and the output of the code execution.
+        """
+        return asyncio.run(self.async_run_code_on_notebook(code))
+
+    def reset_notebook(self) -> ServiceResponse:
+        """
+        Reset the notebook
+        """
+        asyncio.run(self._restart_client())
+        return ServiceResponse(
+            status=ServiceExecStatus.SUCCESS,
+            content="Reset notebook",
+        )
+
+    def _save_image(self, image_base64: str) -> str:
```

Comment:
Save image via `file_manager.save_image`

Reply:
The save image via image_base64 not yet implemented in the `main` branch.

Reply:
Web browser agent need this as well. https://github.com/modelscope/agentscope/blob/ac0e8edf8e9e63c3fa520367bceadd0dc2a8db50/src/agentscope/file_manager.py

Reply:
Using file_manager.save_image